pyne/openmc_utils.py

```diff
@@ -156,3 +171,270 @@ def xml(self):
+    var_tot = np.sum(np.square(std_dev), axis=1)
+    nonzero = res_tot > 0
+    rel_err_tot[nonzero] = np.sqrt(var_tot[nonzero]) / res_tot[nonzero]
+
+    # get volume averged results
+    result = np.divide(result, ve_vol)
+    res_tot = np.divide(res_tot, ve_vol)
+    # In the result of OpenMC, x changes fastest
+    # change the order to z changes fastest
+    result = result_changes_order(result, m.dims[3:6])
+    rel_err = result_changes_order(rel_err, m.dims[3:6])
+    res_tot = result_changes_order(res_tot, m.dims[3:6])
+    rel_err_tot = result_changes_order(rel_err_tot, m.dims[3:6])
+    return result, rel_err, res_tot, rel_err_tot
+
+def result_changes_order(result, dims):
+    """
+    This function changes the order of openmc flux data. The default order of
+    OpenMC flux is x changes the fastest. This function changes it to the order
+    which z changes the fasest (the same order as MCNP meshtal).
+
+    Parameters:
+    -----------
+    result : numpy array
+        This numpy array contains the flux data read from OpenMC state point
+        file. The shape of this numpy array is (num_ves*num_e_groups) or
+        (num_ves, ), with x changes the fastest.
+    dims : list
+        The dimension of x, y, z.
+
+    Returns:
+    --------
+    result_ : numpy array
+        This numpy array contains the flux data read from OpenMC state point
+        file. The shape of this numpy array is (num_ves*num_e_groups) or
+        (num_ves, ), with changes the fastest.
+    """
+
+    result_ = np.zeros_like(result)
+    x_dim, y_dim, z_dim = dims[0], dims[1], dims[2]
+    for i in range(x_dim):
+        for j in range(y_dim):
+            for k in range(z_dim):
+                idx = k * (x_dim * y_dim) + j * x_dim + i
+                idx_ = i * (y_dim * z_dim) + j * z_dim + k
+                result_[idx_] = result[idx]
+    return result_
+
+
+def create_meshtally(filename, tally_id, mesh_id=None, particle=None,
+                     tag_names=None, mesh_has_mats=False):
+    """
+    This function creates a MeshTally instance from OpenMC statepoint file.
+
+    Parameters:
+    -----------
+    filename : str
+        Filename of the OpenMC statepoint file. It ends with ".h5",
+        eg: "statepoint.10.h5".
+    tally_id : int
+        Tally number.
+    mesh_id : int
+        Mesh id of this tally used. Required if multiple meshes exist in the
+        OpenMC state point file.
+    particle : str
+        The particle type, 'neutron' or 'photon'.
+    tag_names : iterable, optional
+        Four strs that specify the tag names for the results, relative
+        errors, total results and relative errors of the total results.
+    mesh_has_mats: bool
+        If false, Meshtally objects will be created without PyNE material
+        objects.
+
+    Returns:
+    --------
+    m : MeshTally object
+        The MeshTally object created from OpenMC state point file with tally
+        number of tally_id.
+    """
+
+    m = MeshTally()
```

Comment:
It might be good to update the `MeshTally.__init__` to take some of these parameters like the `tally_id`, `particle`, `structured_coords`, etc. Right now we create a partially functional object and then fill in additional information as we go. That happens both here and in the `mcnp` module. It’d probably be better to have a fully functional object from the start by collecting all this information in the `create_meshtally` functions and then creating a `MeshTally` instance by passing all necessary information at once.

Could be left to a future PR though.

Reply:
Agreed. That will be in a future PR.